pkg/agent/host_configuration_unix.go

```diff
@@ -0,0 +1,133 @@
+		klog.Errorf("Failed to set host link for %s up: %v", i.hostGateway, err)
+		return err
+	}
+
+	// Configure host gateway IP using the first address of node localSubnet
+	localSubnet := i.nodeConfig.PodCIDR
+	subnetID := localSubnet.IP.Mask(localSubnet.Mask)
+	gwIP := &net.IPNet{IP: ip.NextIP(subnetID), Mask: localSubnet.Mask}
+	gwAddr := &netlink.Addr{IPNet: gwIP, Label: ""}
+	gwMAC := link.Attrs().HardwareAddr
+	i.nodeConfig.GatewayConfig = &types.GatewayConfig{Link: i.hostGateway, IP: gwIP.IP, MAC: gwMAC}
+	gatewayIface.IP = gwIP.IP
+	gatewayIface.MAC = gwMAC
+
+	// Check IP address configuration on existing interface, return if already has target
+	// address
+	// We perform this check unconditionally, even if the OVS port did not exist when this
+	// function was called (i.e. portExists is false). Indeed, it may be possible for the Linux
+	// interface to exist even if the OVS bridge does not exist.
+	if addrs, err := netlink.AddrList(link, netlink.FAMILY_V4); err != nil {
+		klog.Errorf("Failed to query IPv4 address list for interface %s: %v", i.hostGateway, err)
+		return err
+	} else if addrs != nil {
+		for _, addr := range addrs {
+			klog.V(4).Infof("Found IPv4 address %s for interface %s", addr.IP.String(), i.hostGateway)
+			if addr.IP.Equal(gwAddr.IPNet.IP) {
+				klog.V(2).Infof("IPv4 address %s already assigned to interface %s", addr.IP.String(), i.hostGateway)
+				return nil
+			}
+		}
+	} else {
+		klog.V(2).Infof("Link %s has no configured IPv4 address", i.hostGateway)
+	}
+
+	klog.V(2).Infof("Adding address %v to gateway interface %s", gwAddr, i.hostGateway)
+	if err := netlink.AddrAdd(link, gwAddr); err != nil {
+		klog.Errorf("Failed to set gateway interface %s with address %v: %v", i.hostGateway, gwAddr, err)
+		return err
+	}
+
+	// send_redirects for the interface will be enabled if at least one of
+	// conf/{all,interface}/send_redirects is set to TRUE, so "all" and the
+	// interface must be disabled together.
+	// See https://www.kernel.org/doc/Documentation/networking/ip-sysctl.txt.
+	if err := disableICMPSendRedirects("all"); err != nil {
+		return err
+	}
+	if err := disableICMPSendRedirects(i.hostGateway); err != nil {
+		return err
+	}
+	return nil
+}
+
+// prepareHostNetworking returns immediately on Linux.
+func (i *Initializer) prepareHostNetworking() error {
+	return nil
+}
+
+func disableICMPSendRedirects(intfName string) error {
+	cmdStr := fmt.Sprintf("echo 0 > /proc/sys/net/ipv4/conf/%s/send_redirects", intfName)
+	cmd := exec.Command("/bin/sh", "-c", cmdStr)
+	if err := cmd.Run(); err != nil {
+		klog.Errorf("Failed to disable send_redirect for interface %s: %v", intfName, err)
+		return err
+	}
+	return nil
+}
+
+// setupExternalNetworking setups iptables chains and rules.
+func (i *Initializer) setupExternalNetworking() error {
+	// Setup iptables chains and rules.
+	if err := i.iptablesClient.Initialize(i.hostGateway, i.serviceCIDR, i.nodeConfig, i.trafficEncapMode); err != nil {
```

Comment:
I think it is a bit complex if we don't use an interface for iptablesClient. Even if I split the invocations of iptablesClient to a seperate file with "_linux" suffix, we still need iptablesClient as it is a property to construct node_route_controller.Controller which is passed from cmd/agent.go. If we want to remove the interface "iptablesClient", we might differ OS types both in cmd/agent, and pkg/agent/controller/node_route_controller. It makes the change a bit larger.

Reply:
Even we don't want to introduce different platform code in cmd/agent.go. It can just inject a dummy iptables instance to node_route_controller without it being called. What I want to suggest is not calling iptablesClient.AddPeerCIDR and iptablesClient.Reconcile on main path, abstracting an interface or not doesn't matter. 
It was just a reminder that the corresponding functions would then be never called on windows and if you are creating a windows implementation for this purpose, then a dummy iptables struct without any functions can work too.

Reply:
Got it. I will move invocation of iptablesClient in node_route_controller to _unix firstly.